pytoda/datasets/_csv_statistics.py

```diff
@@ -27,40 +33,141 @@ def __init__(
 
         """
         self.filepath = filepath
-        self.feature_list = feature_list
+        self.initial_feature_list = feature_list  # see `setup_datasource`
         self.min_max_scaler = MinMaxScaler()
         self.standardizer = StandardScaler()
         self.kwargs = copy.deepcopy(kwargs)
         self.kwargs['dtype'] = pandas_dtype
-        if self.feature_list is not None:
-            # NOTE: zeros denote missing value
-            self.feature_fn = lambda df: df.T.reindex(self.feature_list
-                                                      ).T.fillna(0.0)
-        else:
-            self.feature_fn = lambda df: df
-        self.feature_mapping = self.setup_datasource()
+
+        self.preprocess_df = (
+            # may be applied to many chunks, so logic is determined once here
+            self._reindex if self.initial_feature_list else self._id
+        )
+        self.setup_datasource()
+
+        # self.notna_count
         self.max = self.min_max_scaler.data_max_
         self.min = self.min_max_scaler.data_min_
         self.mean = self.standardizer.mean_
         self.std = self.standardizer.scale_
 
-    def setup_datasource(self) -> pd.Series:
+    def _reindex(self, df: pd.DataFrame) -> pd.DataFrame:
+        """Missing features are represented with all zeros."""
+        # fill = 0.0
+        # fills with NaN TODO docs
```

Comment:
I think 0-imputation makes sense but this method looks a bit unfinished to me, especially since the default `fill_value` is `np.nan` and not 0.

Reply:
yes good catch, this is a TODO I forgot to clean up.

Reply:
To be clear, we do impute nothing/np.NaN here on purpose.

Reply:
Ok, this is surprising since the old behavior was a 0 imputation. What's the rationale of the nans and what happens with the nan values then during the transformation?

Reply:
At this stage the data is read the first time. Here we need to keep any NaN and fill missing features given by the `initial_feature_list` with NaN (making it explicit here, though it is the default).

This way the statistics can be computed correctly ignoring NaN (this required the changes you commented below e.g. from `np.max` to `np.nanmax`. NaNs stay Nans in the transformation. After transformation the actual imputation is performed.

In other words, the old default of imputing 0 here was wrong.